File: gerar_doc_flows.py
import os
import xml.etree.ElementTree as ET

FLOWS_DIR = r"force-app\main\default\flows"
OUTPUT_FILE = r"Documentacao_Flows.md"
NS = {'ns': 'http://soap.sforce.com/2006/04/metadata'}

def extract_text(element, path):
    if element is None:
        return ""
    node = element.find(f"ns:{path}", NS)
    return node.text if node is not None else ""

def parse_flow(filepath):
    tree = ET.parse(filepath)
    root = tree.getroot()
    
    flow_label = extract_text(root, 'label')
    flow_desc = extract_text(root, 'description')
    
    if not flow_label:
        flow_label = os.path.basename(filepath).split('.')[0]
        
    doc_lines = []
    doc_lines.append(f"## Fluxo: {flow_label}")
    if flow_desc:
        doc_lines.append(f"**Objetivo do Fluxo**: {flow_desc}\n")
    else:
        doc_lines.append(f"**Objetivo do Fluxo**: Fluxo automatizado do Salesforce utilizado no núcleo de negócio.\n")
    
    components = []
    
    def process_node(node, type_name):
        name = extract_text(node, 'name')
        label = extract_text(node, 'label') or name
        desc = extract_text(node, 'description')
        
        comp_info = {"type": type_name, "label": label, "desc": desc, "details": []}
        
        if type_name == "Busca de Registro":
            obj = extract_text(node, 'object')
            comp_info['details'].append(f"**Objeto:** {obj}")
            action = f"Busca e recuperação de dados presentes no objeto {obj} de acordo com as condições definidas."
            objective = desc if desc else f"Extratificar ou identificar o registro do objeto {obj} para uso nos próximos cenários."
            comp_info['action'] = action
            comp_info['objective'] = objective
        elif type_name == "Atribuição de Variável":
            action = "Armazena ou transforma informações atualizando os valores em memória."
            objective = desc if desc else "Preparação ou formatação do pacote de dados utilizado internamente na continuidade do atendimento."
            comp_info['action'] = action
            comp_info['objective'] = objective
        elif type_name == "Decisão Lógica":
            action = "Avalia variáveis da sessão usando regras combinadas e desvia para rotas apropriadas."
            objective = desc if desc else "Identificação lógica do estado (ex: se cliente foi localizado ou não) e definição da experiência em ramificações do cenário."
            comp_info['action'] = action
            comp_info['objective'] = objective
        elif type_name == "Atualização de Registro":
            obj = extract_text(node, 'object')
            comp_info['details'].append(f"**Objeto:** {obj}")
            action = f"Altera pontualmente os dados de um ou mais registros do objeto {obj} em nuvem."
            objective = desc if desc else f"Salvar o estado final da ação atualizando permanentemente as propriedades de {obj} no banco da Org."
            comp_info['action'] = action
            comp_info['objective'] = objective
        elif type_name == "Criação de Registro":
            obj = extract_text(node, 'object')
            comp_info['details'].append(f"**Objeto:** {obj}")
            action = f"Instancia e grava um novo registro único dentro de {obj}."
            objective = desc if desc else f"Processar a criação e vinculação de informações inéditas referentes a {obj} geradas neste fluxo."
            comp_info['action'] = action
            comp_info['objective'] = objective
        elif type_name == "Ação (Action)":
            actionType = extract_text(node, 'actionType')
            action = f"Dispara a capacidade de sistema: {actionType}."
            objective = desc if desc else "Acionar capacidades da plataforma ou de classe invocável complementar à lógica primária."
            comp_info['action'] = action
            comp_info['objective'] = objective
        else:
            comp_info['action'] = "Execução de lógica intrínseca e declarada dentro do ambiente isolado da Salesforce."
            comp_info['objective'] = desc if desc else "Garantir a fluidez, resgate de valor ou continuidade para resolução final."
            
        components.append(comp_info)

    for node in root.findall('ns:start', NS):
        # Para starts, pegamos o objeto se for trigger
        start_obj = extract_text(node, 'object')
        trigger = extract_text(node, 'triggerType')
        if start_obj:
            comp_info = {
                "type": "Gatilho de Início (Start)",
                "label": "Início do Fluxo",
                "desc": "",
                "details": [f"**Objeto do Gatilho:** {start_obj}", f"**Tipo do Gatilho:** {trigger}"]
            }
            comp_info['action'] = "Espera alterações e avalia condições de início disparando de forma proativa o sistema."
            comp_info['objective'] = f"Garantir a iniciação correta da execução para modificações de {start_obj}."
            components.append(comp_info)

    for node in root.findall('ns:recordLookups', NS):
        process_node(node, "Busca de Registro")
        
    for node in root.findall('ns:assignments', NS):
        process_node(node, "Atribuição de Variável")

    for node in root.findall('ns:decisions', NS):
        process_node(node, "Decisão Lógica")

    for node in root.findall('ns:recordUpdates', NS):
        process_node(node, "Atualização de Registro")

    for node in root.findall('ns:recordCreates', NS):
        process_node(node, "Criação de Registro")
        
    for node in root.findall('ns:actionCalls', NS):
        process_node(node, "Ação (Action)")
        
    for comp in components:
        doc_lines.append(f"### {comp['type']}: {comp['label']}")
        if comp['desc']:
            doc_lines.append(f"{comp['desc']}\n")
        doc_lines.append(f"* **Ação:** {comp['action']}")
        doc_lines.append(f"* **Objetivo:** {comp['objective']}")
        for detail in comp['details']:
            doc_lines.append(f"* {detail}")
        doc_lines.append("")

    return "\n".join(doc_lines)

def main():
    root_dir = os.path.abspath(os.path.dirname(__file__))
    flows_abs_dir = os.path.join(root_dir, FLOWS_DIR)
    
    if not os.path.exists(flows_abs_dir):
        print(f"Erro: Diretório {flows_abs_dir} não encontrado.")
        return
        
    output_lines = [
        "# Documentação de Flows do Salesforce",
        "",
        "> Documento gerado de forma automática com base no processamento lógico dos metadados (*.flow-meta.xml*) para centralização da fundação dos dados do ambiente.",
        ""
    ]
    
    files = [f for f in os.listdir(flows_abs_dir) if f.endswith('.flow-meta.xml')]
    files.sort()
    
    for filename in files:
        filepath = os.path.join(flows_abs_dir, filename)
        print(f"Processando: {filename}")
        try:
            doc = parse_flow(filepath)
            output_lines.append(doc)
            output_lines.append("\n---\n")
        except Exception as e:
            print(f"Erro ao processar {filename}: {e}")
            output_lines.append(f"## Fluxo: {filename} (Erro ao ler arquivo XML)\n\n---\n")
            
    out_path = os.path.join(root_dir, OUTPUT_FILE)
    with open(out_path, 'w', encoding='utf-8') as f:
        f.write("\n".join(output_lines))
        
    print(f"Documentado com sucesso {len(files)} arquivos de fluxo! Gerado em: {out_path}")

if __name__ == '__main__':
    main()

File: gerar_doc_flows.js
const fs = require('fs');
const path = require('path');
const businessObjectives = require('./business_objectives.js');

const flowsDir = path.join(__dirname, 'force-app/main/default/flows');
const outputFile = path.join(__dirname, 'Documentacao_Flows.md');

let outputLines = [
    "# Documentação de Flows do Salesforce\n"
];

if (!fs.existsSync(flowsDir)) {
    console.error(`Directory not found: ${flowsDir}`);
    process.exit(1);
}

const files = fs.readdirSync(flowsDir).filter(f => f.endsWith('.flow-meta.xml')).sort();

function extractText(xml, tagName) {
    const regex = new RegExp(`<${tagName}>(.*?)</${tagName}>`, 's');
    const match = xml.match(regex);
    return match ? match[1].trim() : "";
}

function getTarget(xmlChunk) {
    const match = xmlChunk.match(/<connector>.*?<targetReference>(.*?)<\/targetReference>.*?<\/connector>/s);
    return match ? match[1].trim() : null;
}

files.forEach(file => {
    const filePath = path.join(flowsDir, file);
    try {
        const xml = fs.readFileSync(filePath, 'utf8');
        let flowLabel = extractText(xml, 'label') || file.replace('.flow-meta.xml', '');
        let flowDesc = extractText(xml, 'description');
        const apiName = file.replace('.flow-meta.xml', '');
        
        let businessDesc = businessObjectives[apiName] || flowDesc || 'Fluxo automatizado do Salesforce responsável por processar e transacionar eventos de negócios vitais entre o CRM e os endpoints.';
        const processType = extractText(xml, 'processType') || 'Desconhecido';
        
        outputLines.push(`## Fluxo: ${flowLabel}\n`);
        outputLines.push(`**Objetivo do Fluxo**: ${businessDesc}  `);
        outputLines.push(`**Especificações Técnicas Gerais:**\n`);
        outputLines.push(`* **Nome de API:** ${apiName}`);
        
        let typeDesc = "";
        if (processType === 'AutoLaunchedFlow') typeDesc = " (Fluxo iniciado automaticamente)";
        else if (processType === 'Flow') typeDesc = " (Fluxo de Tela)";
        
        outputLines.push(`* **Tipo de Fluxo:** ${processType}${typeDesc}.\n`);

        const typesMap = {
            'recordLookups': 'Busca de Registro',
            'assignments': 'Atribuição de Variável',
            'decisions': 'Decisão Lógica',
            'recordUpdates': 'Atualização de Registro',
            'recordCreates': 'Criação de Registro',
            'actionCalls': 'Ação (Action)'
        };

        let nodesDict = {};
        Object.keys(typesMap).forEach(type => {
            const blockRegex = new RegExp(`<${type}>(.*?)</${type}>`, 'gs');
            let match;
            while ((match = blockRegex.exec(xml)) !== null) {
                let nodeXml = match[1];
                let name = extractText(nodeXml, 'name');
                if (name) {
                    nodesDict[name] = { xml: nodeXml, type: type, typeName: typesMap[type], name: name };
                }
            }
        });

        const startMatch = xml.match(/<start>(.*?)<\/start>/s);
        let startRef = extractText(xml, 'startElementReference');
        
        if (startMatch) {
            const snode = startMatch[1];
            const obj = extractText(snode, 'object');
            if (obj) {
                const trigger = extractText(snode, 'triggerType');
                outputLines.push(`### Gatilho de Início (Start): Entrada do Fluxo\n`);
                outputLines.push(`Ponto de entrada do sistema identificando os eventos ativos da transação.\n`);
                outputLines.push(`* **Ação:** Espera e valida alterações ativando execução automática sob condição Trigger: ${trigger || 'Geral'} sobre a base de ${obj}.`);
                outputLines.push(`* **Objetivo:** Estabelecer arquitetura sistêmica em tempo real e orquestrar metadados da mesa deste processo.\n`);
            }
            if (!startRef) {
                startRef = getTarget(snode);
            }
        }

        let visited = new Set();
        let lastPrintedCondition = "Fluxo Primário (Raiz)";

        function traverse(nodeName, currentCond) {
            if (!nodeName || visited.has(nodeName)) return; // Prevents loop and cleanly handles merges natively
            visited.add(nodeName);

            let nodeInfo = nodesDict[nodeName];
            if (!nodeInfo) return;

            let nx = nodeInfo.xml;
            let label = extractText(nx, 'label') || nodeInfo.name;
            let desc = extractText(nx, 'description');
            let obj = extractText(nx, 'object');
            let actType = extractText(nx, 'actionType');

            // Handling Groupings Visually in Markdown
            if (currentCond !== lastPrintedCondition) {
                outputLines.push(`> 🔀 **TRILHA CONDICIONAL ATIVA:** *${currentCond}*\n`);
                lastPrintedCondition = currentCond;
            }

            outputLines.push(`### ${nodeInfo.typeName}: ${label}\n`);
            if (desc) outputLines.push(`${desc}\n`);
            
            let finalAction = "";
            let finalObj = "";

            if (nodeInfo.type === 'recordLookups') {
                finalAction = `Busca e reestruturação da tabela de base contendo o objeto ${obj || 'relacionado'} usando chaves atreladas à etapa.`;
                finalObj = `Identificar localmente de forma unitária via lookups permitindo que próximos branches herdem as instâncias de ${obj || 'registro'}.`;
            } else if (nodeInfo.type === 'assignments') {
                finalAction = `Atribui ou manipulação local computacional aos arrays ou variáveis da estrutura.`;
                finalObj = `Montar e consolidar os pacotes de dados processados utilizados internamente na continuidade ou devolução.`;
            } else if (nodeInfo.type === 'recordUpdates') {
                finalAction = `Alteração restrita (update/upsert) nos dados da nuvem referenciadas em IDs de ${obj || 'registros'}.`;
                finalObj = `Gravar com persistência absoluta o bloco mutado pela lógica gerando log no repositório.`;
            } else if (nodeInfo.type === 'recordCreates') {
                finalAction = `Inicia geração transacional do tipo ${obj || 'registro'} preenchendo todos as propriedades mapeadas no decorrer do flow.`;
                finalObj = `Registrar existencialmente para rastreamento físico criando ligações orgânicas em banco de dados.`;
            } else if (nodeInfo.type === 'actionCalls') {
                finalAction = `Dispara o endpoint de capacidades restritas e externas mapeado por ${actType || 'Action'}.`;
                finalObj = `Acionar comportamentos e engenharias assíncronas do negócio que escapam o simples manipulamento de variaveis locais.`;
            }

            if (nodeInfo.type === 'decisions') {
                if (!desc) outputLines.push(`O sistema toma a bifurcação de caminho baseando-se em lógicas booleanas exclusivas calculadas.\n`);
                
                let rulesRegex = /<rules>(.*?)<\/rules>/gs;
                let rm;
                let ruleIdx = 1;

                while ((rm = rulesRegex.exec(nx)) !== null) {
                    let rx = rm[1];
                    let rLabel = extractText(rx, 'label') || `Cenário ${ruleIdx}`;
                    outputLines.push(`* **Cenário ${ruleIdx} (${rLabel}):** Assumirá a execução atrelada a esta lógica em caso próspero.`);
                    
                    let target = getTarget(rx);
                    if (target) {
                        outputLines.push(""); // spacer
                        traverse(target, `Cenário do(a) '${label}': ${rLabel}`);
                    }
                    ruleIdx++;
                }

                let dConnMatch = nx.match(/<defaultConnector>.*?<targetReference>(.*?)<\/targetReference>.*?<\/defaultConnector>/s);
                let dTarget = dConnMatch ? dConnMatch[1].trim() : null;
                let dLabel = extractText(nx, 'defaultConnectorLabel') || 'Ponto de Partida Falho / Padrão';
                outputLines.push(`* **Cenário ${ruleIdx} (${dLabel}):** Via fallback o sistema executa esta rota em caso de invalidações nas validações centrais.`);
                if (dTarget) {
                    outputLines.push("");
                    traverse(dTarget, `Cenário Padrão do(a) '${label}': ${dLabel}`);
                }
            } else {
                outputLines.push(`* **Ação:** ${finalAction}`);
                outputLines.push(`* **Objetivo:** ${finalObj}\n`);

                let target = getTarget(nx);
                if (target) {
                    traverse(target, currentCond);
                }
            }
        }

        if (startRef) {
            traverse(startRef, "Fluxo Principal");
        }

        // Failsafe para nós desconectados graficamente 
        Object.keys(nodesDict).forEach((k) => {
            if (!visited.has(k)) {
                traverse(k, "Trilha Desconectada / Fluxos Assíncronos / Orfãos");
            }
        });

        outputLines.push("---\n");

    } catch (e) {
        outputLines.push(`## Fluxo: ${file} (Erro ao processar Grafo)\n\n---\n`);
    }
});

fs.writeFileSync(outputFile, outputLines.join('\n').trim() + '\n', 'utf8');
console.log('Documentacao_Flows.md was grouped conceptually via DFS successfully in UTF-8.');
